File: go/stream_shm.go
package main

import (
	"fmt"
	"os"
	"syscall"
	"unsafe"
)

const (
	SHM_NAME       = "/dev/shm/msg_queue"
	MAX_FLOAT_SIZE = 16
	MAX_QUEUE_SIZE = 100000
	MSG_SIZE       = int(unsafe.Sizeof(Msg{}))
	SHM_SIZE       = int(unsafe.Sizeof(Queue{}))
)

type Msg struct {
	// 1: Binance-Futures_BTCUSDT, 2: Binance-Futures_ETHUSDT, 3: Binance-Futures_SOLUSDT, 4: Binance-Futures_DOGEUSDT, 5: Binance_BTCUSDT
	InstrumentID int64
	// 1: L1 Bid, -1: L1 Ask, 2: L2 Bid, -2: L2 Ask, 3: Buy Trade, -3: Sell Trade
	MsgType int64
	// Transaction Time MS
	TxMs int64
	// Event Time MS
	EventMs int64
	// Local Time NS
	LocalNs int64
	// Sequence Number / Trade ID
	SnID int64
	// Price in String
	Price [MAX_FLOAT_SIZE]byte
	// Size in String
	Size [MAX_FLOAT_SIZE]byte
}

type Queue struct {
	Sn   int64
	From int64
	To   int64
	Msgs [MAX_QUEUE_SIZE * MSG_SIZE]byte
}

func main() {
	// 打開共享記憶體區域
	fd, err := syscall.Open(SHM_NAME, syscall.O_RDWR, 0666)
	if err != nil {
		fmt.Fprintf(os.Stderr, "shm_open error: %v\n", err)
		os.Exit(1)
	}
	defer syscall.Close(fd)

	// 映射共享記憶體
	ptr, err := syscall.Mmap(fd, 0, SHM_SIZE, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED)
	if err != nil {
		fmt.Fprintf(os.Stderr, "mmap error: %v\n", err)
		os.Exit(1)
	}
	defer syscall.Munmap(ptr)

	// 將共享記憶體的指針轉換為 Queue 結構
	buf := (*Queue)(unsafe.Pointer(&ptr[0]))
	cur := buf.From

	// 循環讀取共享記憶體中的消息
	for {
		for cur != buf.To {
			// 計算當前消息的起始地址
			msgPtr := (*Msg)(unsafe.Pointer(&buf.Msgs[cur*int64(MSG_SIZE)]))

			// 打印消息內容
			fmt.Printf("%d: %d, %d, %d, %d, %d, %d, %s, %s\n",
				cur,
				msgPtr.InstrumentID,
				msgPtr.MsgType,
				msgPtr.SnID,
				msgPtr.TxMs,
				msgPtr.EventMs,
				msgPtr.LocalNs,
				string(msgPtr.Price[:]),
				string(msgPtr.Size[:]))

			// 更新讀取位置
			cur = (cur + 1) % MAX_QUEUE_SIZE
		}
	}
}

File: c/stream_shm.c
#define SHM_NAME "/msg_queue"
#define MAX_FLOAT_SIZE 16
#define MAX_QUEUE_SIZE 100000
#define MSG_SIZE sizeof(Msg)
#define SHM_SIZE sizeof(Queue)

typedef struct Msg
{
    // 1: Binance-Futures_BTCUSDT, 2: Binance-Futures_ETHUSDT, 3: Binance-Futures_SOLUSDT, 4: Binance-Futures_DOGEUSDT, 5: Binance_BTCUSDT
    long instrument_id;
    // 1: L1 Bid, -1: L1 Ask, 2: L2 Bid, -2: L2 Ask, 3: Buy Trade, -3: Sell Trade
    long msg_type;
    // Transaction Time MS
    long tx_ms;
    // Event Time MS
    long event_ms;
    // Local Time NS
    long local_ns;
    // Sequence Number / Trade ID
    long sn_id;
    // Price
    char price[MAX_FLOAT_SIZE];
    // Size
    char size[MAX_FLOAT_SIZE];
} Msg;

typedef struct
{
    long sn;
    long from;
    long to;
    char msgs[MAX_QUEUE_SIZE * MSG_SIZE];
} Queue;

// main.c
#include <fcntl.h>
#include <sys/mman.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>

int fd = 0;
void *ptr;
Queue *buf;

int main()
{
    // 打開共享記憶體區域
    fd = shm_open(SHM_NAME, O_RDWR, 0666);
    if (fd == -1)
    {
        perror("shm_open");
        exit(1);
    }

    // 映射共享記憶體
    ptr = mmap(0, SHM_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    if (ptr == MAP_FAILED)
    {
        perror("mmap");
        close(fd);
        exit(1);
    }

    // 讀取共享記憶體中的 Buf 結構體數據
    buf = (Queue *)ptr;
    long cur = buf->from;

    while (1)
    {
        while (cur != buf->to)
        {
            Msg *msg = (Msg *)&buf->msgs[cur * MSG_SIZE];
            printf("%lli: %lli, %lli, %lli, %lli, %lli, %lli, %s, %s\n", cur, msg->instrument_id, msg->msg_type, msg->sn_id, msg->tx_ms, msg->event_ms, msg->local_ns, msg->price, msg->size);
            cur = (cur + 1) % MAX_QUEUE_SIZE;
        }
    }

    // 解除映射並關閉共享記憶體
    munmap(ptr, SHM_SIZE);
    close(fd);

    return 0;
}
